[Merged by Bors] - feat: add ComplexShape.EulerCharSigns for generalized Euler characteristic

jessealama opened the pull request:
Add `ComplexShape.EulerCharSigns` typeclass providing the alternating signs `χ : ι → ℤˣ` for Euler characteristic computations. Instances are provided for `up ℕ`, `down ℕ`, `up ℤ`, and `down ℤ`.

The Euler characteristic definition (`eulerChar`) uses `finsum` to sum over all indices, with the `ComplexShape` as an explicit parameter. This allows working with arbitrary index types without requiring a `Fintype` instance, defaulting to 0 when the support is infinite.

The `GradedObject` version is the primary definition, with `HomologicalComplex.eulerChar` and `HomologicalComplex.homologyEulerChar` as abbreviations that apply the graded object version to `C.X` and `C.homology` respectively.

Split from #29713 as suggested by @joelriou.

joelriou commented on Mathlib/Algebra/Homology/EulerCharacteristic.lean:
```diff
@@ -0,0 +1,150 @@
+
+/-- If homology vanishes outside a finite set, the infinite homological Euler
+characteristic equals the finite one. -/
+theorem homologyEulerCharTsum_eq_homologyEulerChar [c.EulerCharSigns]
```
Could you introduce definitions similar to `eulerChar` and `eulerCharTsum` but for `GradedObject`? (This would require adding `c` as an explicit parameter.)
Then, you may introduce abbrevs for these two definitions in the particular case of `C.X` and `C.homology`.
The lemma `homologyEulerCharTsum_eq_homologyEulerChar` should be stated for this more general definition on graded objects.

jessealama replied:
Great idea! Done.

jessealama commented:
> This pull request has conflicts, please merge `master` and resolve them.

resolved!

jessealama committed:
feat: add ComplexShape.EulerCharSigns for generalized Euler characteristic

Add Euler characteristic for homological complexes with arbitrary index types
and shapes. The typeclass ComplexShape.EulerCharSigns provides alternating
signs for Euler characteristic computations, with instances for up/down N and
up/down Z.

Introduces GradedObject.eulerChar and GradedObject.eulerCharTsum with
ComplexShape as an explicit parameter, with HomologicalComplex versions as
abbreviations.

joelriou commented:
Could you run `lake exe mk_all` because it seems your `Mathlib.lean` file is broken?

robin-carlier commented on Mathlib/Algebra/Homology/EulerCharacteristic.lean:
```diff
@@ -0,0 +1,187 @@
+
+variable [Fintype ι]
+
+/-- The Euler characteristic as an infinite sum over all indices.
```
```suggestion
/-- The Euler characteristic as a sum over all indices.
```

robin-carlier reviewed (commented):
The biggest problem here is that the definitions of the "infinite" euler characteristics are misleading, as they take a Fintype argument. As a result, all the theorems that mentions them is in fact asking the index type for the graded object to be finite.

jessealama committed:
Update Mathlib/Algebra/Homology/EulerCharacteristic.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

jessealama committed:
feat(Homology): use finsum for eulerCharTsum, remove Fintype requirement

jessealama committed:
feat(Function.support): add support_eq_univ and support_mul_of_ne_zero lemmas

Add `mulSupport_eq_univ`/`support_eq_univ` for functions that are everywhere
non-one/nonzero, and `support_mul_of_ne_zero_left/right` for products where
one factor is everywhere nonzero.

Simplify `support_eulerChar_summand` using the new lemmas.

robin-carlier reviewed (commented):
Now that the definition of `eulerCharTSum` (which should be named differently, given that it’s not directly a `tsum` in the sense of [tsum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/InfiniteSum/Defs.html#tsum)) is general enough as a `finsum`, is there much point in keeping the current `eulerChar` as a definition on its own? I think almost always the `finsum` is going to be the one we care about mathematically, so it’s probably the one that deserves the name of `eulerChar`.

The lemma that we can compute it as a sum over any finset containing the support is still valuable IMO, but the lemma doesn’t have to be stated with a separate `def` for the sum in question. So I think we could directly go to

```lean
/-- The support of a graded object with respect to finite rank:
the set of indices where the rank is nonzero. -/
def finrankSupport (X : CategoryTheory.GradedObject ι (ModuleCat R)) : Set ι :=
  Function.support (fun i => Module.finrank R (X i))

/-- The finite rank support is contained in a set if and only if
the rank vanishes outside that set. -/
lemma finrankSupport_subset_iff (X : CategoryTheory.GradedObject ι (ModuleCat R)) (s : Set ι) :
    finrankSupport X ⊆ s ↔ ∀ i ∉ s, Module.finrank R (X i) = 0 :=
  Function.support_subset_iff'

/-- The Euler characteristic as a sum over all indices. Defaults to 0 if the support of the ranks
of the objects of `X` is infinite. -/
noncomputable def eulerChar (X : CategoryTheory.GradedObject ι (ModuleCat R)) : ℤ :=
  ∑ᶠ i : ι, (c.χ i : ℤ) * Module.finrank R (X i)

/-- The support of the Euler characteristic summand equals the finite rank support,
because the sign `c.χ i` is always nonzero (it's a unit). -/
private lemma support_eulerChar_summand (X : CategoryTheory.GradedObject ι (ModuleCat R)) :
    Function.support (fun i => (c.χ i : ℤ) * Module.finrank R (X i)) = finrankSupport X := by
  simp only [finrankSupport, Function.support_mul_of_ne_zero_left (fun i => Units.ne_zero (c.χ i))]
  ext i; simp [Function.mem_support]

/-- The euler characteristic of of a graded object can be computed as the alternating
sum over any finite set containing `finrankSupport X`. -/
theorem eulerChar_eq_sum_finSet_of_finrankSupport_subset (X : CategoryTheory.GradedObject ι (ModuleCat R))
    (indices : Finset ι) (h_support : finrankSupport X ⊆ indices) :
    eulerChar c X = ∑ i ∈ indices, (c.χ i : ℤ) * Module.finrank R (X i) := by
  simp only [eulerChar]
  rw [finsum_eq_sum_of_support_subset]
  exact (support_eulerChar_summand c X).symm ▸ h_support
```

Note that I made the lemma about `support_eulerChar_summand` private: I don’t think the lemma is that interesting or 
useful once you have the theorem `eulerChar_eq_sum_finSet_of_finrankSupport_subset`.

An other important point is documentation of junk values: the one coming from the fact that `finsum` equals `0` if the set is infinite is already documented here, but recall that also `Module.finrank` takes `0` as a junk value if e.g the module is not finitely generated. The `finrankSupport` might be "finite-looking" for complexes that are in fact full of infinite-dimensional components.
I am not saying that this is a bad thing, but this is something that should be clearly documented in the definitions and in the module docstring, because they are easily forgotten and are easy sources of confusion.

jessealama committed:
refactor(Homology): simplify Euler characteristic API

Make finsum-based eulerChar the primary definition, remove separate
finite-sum versions. Make support_eulerChar_summand private.

Thanks @robin-carlier for the review feedback.

jessealama committed:
docs(Homology): clarify junk value behavior in EulerCharacteristic

Add a "Junk values" section to document that `eulerChar` and related
definitions may produce junk values from `finsum` (0 for infinite support)
and `Module.finrank` (0 for modules not free of finite rank).

jessealama commented:
-awaiting-author

jessealama commented on Mathlib/Algebra/Homology/EulerCharacteristic.lean:
```diff
@@ -0,0 +1,176 @@
+public import Mathlib.CategoryTheory.GradedObject
+public import Mathlib.LinearAlgebra.Dimension.Finrank
+public import Mathlib.Algebra.BigOperators.Finprod
+public import Mathlib.Algebra.GroupWithZero.Indicator
```
Ah, nice -- I didn't know about `#min_imports`. Thanks!

jessealama committed:
Update Mathlib/Algebra/Homology/EulerCharacteristic.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

jessealama committed:
Update Mathlib/Algebra/Homology/EulerCharacteristic.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

robin-carlier reviewed (commented):
Thanks!

maintainer merge

jessealama commented:
> Thanks!

Thanks for the help!
